src/ipv6-pinhole-test.cpp

```diff
@@ -0,0 +1,322 @@
+constexpr uint8_t PCP_OPTION_PREFER_FAILURE = 2;
+//! Request header size in bytes (see section 7.1).
+constexpr size_t PCP_REQUEST_HDR_SIZE = 24;
+//! Response header size in bytes (see section 7.2).
+constexpr size_t PCP_RESPONSE_HDR_SIZE = 24;
+//! Option header size in bytes (see section 7.2).
+constexpr size_t PCP_OPTION_HDR_SIZE = 4;
+//! Map request size in bytes (see section 11.1).
+constexpr size_t PCP_MAP_REQUEST_SIZE = 36;
+//! Map response size in bytes (see section 11.1).
+constexpr size_t PCP_MAP_RESPONSE_SIZE = 36;
+//! Mapping nonce size in bytes (see section 11.1).
+constexpr size_t PCP_MAP_NONCE_SIZE = 12;
+//! Result code representing SUCCESS status (7.4).
+constexpr uint8_t PCP_RESULT_SUCCESS = 0;
+
+//! Find IPv6 default gateway.
+std::optional<sockaddr_in6> FindIPv6DefaultGateway()
+{
+    const auto& [read_ok, data] = ReadBinaryFile("/proc/net/ipv6_route");
+    if (!read_ok) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp6: Could not read /proc/net/ipv6_route to find default IPv6 gateway\n");
+        return std::nullopt;
+    }
+    // 149 length (longer if interface name is longer)
+    // destination                      prefix                                 nexthop                                                     flags     iface
+    // 00000000000000000000000000000000 00 00000000000000000000000000000000 00 fe80000000000000da58d7fffe0077cd 00000064 00000007 00000000 00000003     eth0
+    for(std::string &line: SplitString(data, "\n")) {
+        if (line.size() < 149) continue;
+        const std::optional<std::vector<uint8_t>> nexthop(TryParseHex<uint8_t>(line.substr(72, 32)));
+        const std::optional<uint8_t> prefix(ToIntegral<uint8_t>(line.substr(33,  2), ADDR_IPV6_SIZE));
+        const std::optional<uint32_t> flags(ToIntegral<uint32_t>(line.substr(132,  8), ADDR_IPV6_SIZE));
+        const std::string iface(TrimString(line.substr(141)));
+        if (!prefix || !nexthop | !flags || nexthop->size() != ADDR_IPV6_SIZE || iface.empty()) continue; // Parse error
+        if (prefix.value() != 0 || (flags.value() & (RTF_GATEWAY | RTF_UP)) != (RTF_GATEWAY | RTF_UP)) continue; // Not default gw
+        // Look up scope id for interface name string.
+        int gateway_scope_id = if_nametoindex(iface.c_str());
+        if (gateway_scope_id <= 0) {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp6: Could not parse gateway interface %s\n", iface.c_str());
+            return std::nullopt;
+        }
+
+        sockaddr_in6 dest_addr = {};
+        dest_addr.sin6_family = AF_INET6;
+        memcpy(&dest_addr.sin6_addr, nexthop->data(), ADDR_IPV6_SIZE);
+        dest_addr.sin6_scope_id = gateway_scope_id;
+        return dest_addr;
+    }
+    return std::nullopt;
+}
+
+//! Try to open a IPv6 pinhole port using RFC 6887 Port Control Protocol (PCP).
+bool OpenPinhole(const sockaddr_in6 &gateway, const struct in6_addr &my_addr, uint16_t port, uint32_t lifetime, int num_tries = 3)
```

Comment:
IIUC this might as well be called `RequestPortMap`, with the clarification that this boils down to a pinhole for IPv6.